model/supervised_finetuning/custom_datasets/dialogue_collator.py

```diff
@@ -76,3 +77,108 @@ def __call__(self, features):
         batch["targets"] = torch.roll(batch["input_ids"], -1, -1)
 
         return batch
+
+
+@dataclass
+class TrainDialogueDataCollator:
+    """
+    Expects a list of texts corresponding to a sequence of [question, answer, question, answer, ...] pairs.
+    """
+
+    tokenizer: PreTrainedTokenizerBase
+    padding: Union[bool, str, PaddingStrategy] = True
+    max_length: Optional[int] = None
+    mix_length_threshold: Optional[int] = 256
+    mix_probability: Optional[int] = 0.6
+    pad_to_multiple_of: Optional[int] = None
+
+    def __call__(self, features):
+        flatten_messages = []
+        label_masks = []
+        total_short_context = 0
+        for messages in features:
+            messages = list(messages)
+
+            # Add a way for the model to terminate generation
+            # When we predict the start of a new expected question, we want to be able to stop generation
+            messages.append(self.tokenizer.eos_token)
```

Comment:
For this collator to work, we need to replace this eos token with the <Question> tag. Let me know if I am wrong.

Reply:
I would change it anyway, for both this collator and the default collator